src/service.rs

```diff
@@ -263,33 +263,118 @@ fn build_bitbucket_url(user: &str, repo: &str, cfg: &Config, page: &Page) -> Res
     }
 }
 
+fn build_azdevops_url(team: &str, repo: &str, cfg: &Config, page: &Page) -> Result<String> {
+    match page {
+        Page::Open {
+            pull_request: true, ..
+        } => {
+            if let Some(ref b) = cfg.branch {
+                Ok(format!("https://dev.azure.com/{}/_git/{}/pullrequestcreate?sourceRef={}&targetRef=master", team, repo, b))
+            } else {
+                Err(Error::NoLocalRepoFound {
+                    operation: "opening a pull request without specifying branch".to_string(),
+                })
+            }
+        }
+
+        Page::Open { .. } => {
+            if let Some(ref b) = cfg.branch {
+                Ok(format!(
+                    "https://dev.azure.com/{}/_git/{}?version=GB{}",
+                    team, repo, b
+                ))
+            } else {
+                Ok(format!("https://dev.azure.com/{}/{}", team, repo))
+            }
+        }
+
+        Page::Commit { ref hash } => Ok(format!(
+            "https://dev.azure.com/{}/_git/{}/commit/{}",
+            team, repo, hash
+        )),
+
+        Page::Tag { ref tagname, .. } => Ok(format!(
+            "https://dev.azure.com/{}/_git/{}?version=GT{}",
+            team, repo, tagname
+        )),
+
+        Page::FileOrDir {
+            ref relative_path,
+            ref hash,
+            line: None,
+            blame,
+        } => Ok(format!(
+            "https://dev.azure.com/{}/_git/{}/commit/{}?path={}{}",
+            team,
+            repo,
+            hash,
+            to_slash(relative_path),
+            if *blame { "?_a=annotate" } else { "" },
+        )),
+
+        Page::Issue { number } => Ok(format!(
+            "https://dev.azure.com/{}/{}/_workitems/edit/{}",
+            team, repo, number
+        )),
+
+        _ => Err(Error::AzureDevOpsNotSupported),
+    }
+}
+
 // Note: Parse '/user/repo.git' or '/user/repo' or 'user/repo' into 'user' and 'repo'
 pub fn slug_from_path<'a>(path: &'a str) -> Result<(&'a str, &'a str)> {
     let mut split = path.split('/').skip_while(|s| s.is_empty());
-    let user = split.next().ok_or_else(|| Error::NoUserInPath {
+    let mut user = split.next().ok_or_else(|| Error::NoUserInPath {
         path: path.to_string(),
     })?;
+
+    if user == "v3-special-az-devops-case" {
+        // Skip v3 in azure devops ssh urls.
+        user = split.next().ok_or_else(|| Error::NoUserInPath {
+            path: path.to_string(),
+        })?;
+    }
+
     let mut repo = split.next().ok_or_else(|| Error::NoRepoInPath {
         path: path.to_string(),
     })?;
     if repo.ends_with(".git") {
         // Slice '.git' from 'repo.git'
         repo = &repo[0..repo.len() - 4];
+    } else if repo.ends_with("_git") {
+        // Handle special case for weird azure urls.
+        repo = split.next().ok_or_else(|| Error::NoRepoInPath {
+            path: path.to_string(),
+        })?;
```

Comment:
This would make an issue when repository name on GitHub ends with '_git'.

I think it would make code simpler to have another function `azule_dev_ops_slug_from_path`. `build_page_url` function can call  `slug_from_path` or `azule_dev_ops_slug_from_path` looking host.

Reply:
done